README.md
# Wit Node.js SDK [![npm](https://img.shields.io/npm/v/node-wit.svg)](https://www.npmjs.com/package/node-wit)

`node-wit` is the Node.js SDK for [Wit.ai](https://wit.ai).

## Install

In your Node.js project, run:

```bash
npm install --save node-wit
```

## Quickstart

Run in your terminal:

```bash
node examples/basic.js <WIT_TOKEN>
```

See `examples` folder for more examples. Some examples have associated .zip files, do not forget to import those [when creating a new app](https://wit.ai/apps) and grab your access token from the Settings section.

### Messenger integration example

See `examples/messenger.js` for a thoroughly documented tutorial.

### Overview

The Wit module provides a Wit class with the following methods:

- `runComposerAudio` - the [Composer](https://wit.ai/docs/recipes#composer) integration for voice;
- `runComposer` - the [Composer](https://wit.ai/docs/recipes#composer) integration for other inputs;
- `converse` - the Wit [converse](https://wit.ai/docs/http/#post__converse_link) API;
- `event` - the Wit [event](https://wit.ai/docs/http#post__event_link) API;
- `message` - the Wit [message](https://wit.ai/docs/http#get__message_link) API;
- `speech` - the Wit [speech](https://wit.ai/docs/http#post__speech_link) API;
- `dictation` - the Wit [dictation](https://wit.ai/docs/http#post__dictation_link) API;
- `synthesize` - the Wit [synthesize](https://wit.ai/docs/http#post__synthesize_link) API.

You can also require a library function to test out your Wit app in the terminal. `require('node-wit').interactive`

### Wit class

The Wit constructor takes the following parameters:

- `accessToken` - the access token of your Wit instance;
- `actions` - the object of [client action definitions for Composer](https://wit.ai/docs/recipes#run-custom-code);
- `logger` - (optional) the object handling the logging;
- `apiVersion` - (optional) the API version to use instead of the recommended one

The `logger` object should implement the methods `debug`, `info`, `warn` and `error`.
They can receive an arbitrary number of parameters to log.
For convenience, we provide a `Logger` class, taking a log level parameter

Example:

```js
const {Wit, log} = require('node-wit');

const actions = {
  confirm_order(contextMap) {
    return {context_map: {...contextMap, order_confirmation: 'PIZZA42'}};
  },
};

const client = new Wit({
  accessToken: MY_TOKEN,
  actions,
  logger: new log.Logger(log.DEBUG), // optional
});

console.log(client.message('set an alarm tomorrow at 7am'));
```

## .runComposerAudio()

The [Composer](https://wit.ai/docs/recipes#composer) integration for voice.

Takes the following parameters:

- `sessionId` - a unique string identifying the user session
- `contentType` - the Content-Type header
- `body` - the audio `Readable` stream
- `contextMap` - the [context map](https://wit.ai/docs/recipes#custom-context) object

Emits `partialTranscription`, `response` and `fullTranscription` events.
Run the provided `actions` as instructed by the API response, and calls back with the resulting updated context map (unless the action returns `stop: true`).
The Promise returns the final JSON payload of the last API call ([POST /converse](https://wit.ai/docs/http#post__converse_link) or [POST
/event](https://wit.ai/docs/http#post__event_link)).

See `lib/interactive.js` for an example.

## .runComposer()

The [Composer](https://wit.ai/docs/recipes#composer) integration for other
inputs, including text.

Takes the following parameters:

- `sessionId` - a unique string identifying the user session
- `contextMap` - the [context map](https://wit.ai/docs/recipes#custom-context) object
- `message` - the optional user text query

Emits `response` events.
Run the provided `actions` as instructed by the API response, and calls back with the resulting updated context map (unless the action returns `stop: true`).
The Promise returns the final JSON payload of the last [POST /event](https://wit.ai/docs/http#post__event_link) API call.

See `lib/interactive.js` for an example.

## .converse()

The Wit [converse](https://wit.ai/docs/http/#post__converse_link) API.

Takes the following parameters:

- `sessionId` - a unique string identifying the user session
- `contentType` - the Content-Type header
- `body` - the audio `Readable` stream
- `contextMap` - the [context map](https://wit.ai/docs/recipes#custom-context) object

Emits `partialTranscription`, `fullTranscription`, and `response` events. Runs
intermediate `actions` as instructed by the API.

We recommend to use `.runComposerAudio()` instead of this raw API.

## .event()

The Wit [event](https://wit.ai/docs/http#post__event_link) API.

Takes the following parameters:

- `sessionId` - a unique string identifying the user session
- `contextMap` - the [context map](https://wit.ai/docs/recipes#custom-context) object
- `message` - the optional user text query

Emits `response` events, and run intermediate `actions` as instructed by the API.

We recommend to use `.runComposer()` instead of this raw API.

### .message()

The Wit [message](https://wit.ai/docs/http/#get__message_link) API.

Takes the following parameters:

- `q` - the text input you want Wit.ai to extract the information from
- `context` - (optional) the [Context](https://wit.ai/docs/http/#context_link) object
- `n` - (optional) the max number of intents and traits to get back

Example:

```js
const client = new Wit({accessToken: 'MY_TOKEN'});
client
  .message('what is the weather in London?', {})
  .then(data => {
    console.log('Yay, got Wit.ai response: ' + JSON.stringify(data));
  })
  .catch(console.error);
```

See `lib/interactive.js` for another example integration.

### .speech()

The Wit [speech](https://wit.ai/docs/http#post__speech_link) API.

Takes the following paramters:

- `contentType` - the Content-Type header
- `body` - the audio `Readable` stream
- `context` - (optional) the [Context](https://wit.ai/docs/http/#context_link) object
- `n` - (optional) the max number of intents and traits to get back

Emits `partialTranscription`, `partialUnderstanding` and `fullTranscription` events.
The Promise returns the final JSON payload.

See `lib/interactive.js` for an example.

### .dictation()

The Wit [dictation](https://wit.ai/docs/http#post__dictation_link) API.

Takes the following paramters:

- `contentType` - the Content-Type header
- `body` - the audio `Readable` stream

Emits `partialTranscription`, and `fullTranscription` events.
The Promise returns the final JSON payload.

See `examples/synthesize-speech.js` for an example.

### .synthesize()

The Wit [synthesize](https://wit.ai/docs/http#post__synthesize_link) API.

Takes the following paramters (click on link above for more details):

- `q` - The query containting text to synthesize
- `voice` - The voice name. For voices and styles available, see GET [voices.](https://wit.ai/docs/http#get__voices_link)
- `style` - (optional) The style to speak in
- `speed` - (optional) the speed the text is spoken
- `pitch` - (optional) the pitch of the audio
- `gain` - (optional) the gain of the audio

The Promise returns the final response, with the body containing the audio stream of the synthesized text.

See `examples/synthesize-speech.js` for an example.

### interactive

Starts an interactive conversation with your Wit app.

Full conversational interactions:
Use `!converse` to send an audio request from the microphone using Composer.
Enter any text input to send a text request using Composer.

One-off natural language requests:
Use `!speech` to send an audio request from the microphone.
Use `!message <your message>` to send a text request.

Example:

```js
const {interactive} = require('node-wit');
interactive(client);
```

See the [docs](https://wit.ai/docs) for more information.

## Changing the API version

The default (recommended, latest) API version is set in `config.js`.

On May 13th, 2020, the `GET /message` API was updated to reflect the new data model: intents, traits and entities are now distinct.
You can target a specific version by passing the `apiVersion` parameter when creating the `Wit` object.

```json
{
  "text": "hello",
  "intents": [
    {
      "id": "1353535345345",
      "name": "greet",
      "confidence": 0.9753
    }
  ],
  "entities": [],
  "traits": []
}
```

## Running tests

1. Create a new app in wit.ai web console using tests/wit-ai-basic-app-for-tests.zip
2. Copy the Server Access Token from app settings
3. Run `WIT_TOKEN=XXX npm test`, where XXX is the Server Access Token

## License

The license for node-wit can be found in LICENSE file in the root directory of
this source tree.

## Terms of Use
Our terms of use can be found at https://opensource.facebook.com/legal/terms.

Use of Wit.ai services fall under the terms of use found here: https://wit.ai/terms.

## Privacy Policy
Our privacy policy can be found at https://opensource.facebook.com/legal/privacy.

The privacy policy for the Wit.ai service can be found at https://wit.ai/privacy.


CHANGES.md
## v6.6.0 - Multi-responses in Composer

- Multi-responses in Composer: `runComposerAudio()`, `.runComposer()` (and raw `.converse()` + `.event()`) now emit `response` events for intermediate responses, and run intermediate actions as well.
- Bumped API version to `20230215`.

## v6.5.1
- Update uuid to version 9.0.0

## v6.5.0 - Composer alpha integration

- Composer integration: `.runComposerAudio()`, `.runComposer()` (and raw `.converse()` + `.event()`) + `actions` support
- Bumped API version to `20220801`.
- interactive now uses Composer for text inputs, use `!message` for `GET /message` and `!converse` for Composer audio inputs
- added pizza example

## v6.4.0

- Add `POST /synthesize` integration.
- Add `POST /dictation` integration.
- New example using `synthesize()` and `dictation()`.

## v6.3.0

- `speech()` emits `partialUnderstanding` events to support live understanding.
- `apiVersion` updated to `20220608` and its type is now a number.

## Breaking changes

- Bumped API version to `20220608`.
- Emits `partialUnderstanding` events (live understanding support).
- Updated `apiVersion` type from string to number.

## v6.2.2

- Fixes parsing for large HTTP chunks.

## v6.2.1

- Emits `partialTranscription` and `fullTranscription` events.
- Shows microphone input feedback for `interactive`.
- Includes `proxy` support for `speech()`.

## v6.2.0

Requires Node.js >= 6.17.1 to support ES6 directly.

## v6.1.1

- Basic `POST /speech` integration.
- `!speech` support for interactive.

## v6.1.0

Bumped API version to `20210928`.
Moved API version from `Accept` header to `v` HTTP parameter.
Kaizens.

## v6.0.1

Removed unused `request` dependency
Updated various dependencies.

## v6.0.0

Updated API version to latest: `20200513`.
Browse the latest HTTP API documentation [here](https://wit.ai/docs/http/20200513#get__message_link).

## v5.0.0

The most important change is the removal of `.converse()` and `.runActions()`. Follow the migration tutorial [here](https://github.com/wit-ai/wit-stories-migration-tutorial), or [read more here](https://wit.ai/blog/2017/07/27/sunsetting-stories).

### Breaking changes

- `converse` and `runActions` are removed
- updated and added new examples that leverage the /message API
- updated wit-ai-basic-app-for-tests.zip for testing

## v4.3.0

- `converse` and `runActions` are deprecated
- `interactive` now calls `message`

## v4.2.0

- support actions that do not return promises
- support the case where an action does not return a Promise
- update uuid to version 3.0.0
- Support older versions of node
- 'Use strict' on interactive.js
- Check for bot's message in messenger example

## v4.1.0

- Support for different JS environments
- `converse` now takes `reset` as an optional parameter

### Breaking changes

- `interactive` is no longer a function on the `Wit` client. Instead, you require it from the library: `require('node-wit').interactive`
- `runActions` now resets the last turn on new messages and errors.

## v4.0.0

After a lot of internal dogfooding and bot building, we decided to change the API in a backwards-incompatible way. The changes are described below and aim to simplify user code and accommodate upcoming features.

We moved to a Promise-based API, instead of callbacks. This makes the code simpler and the error-handling more straight-forward. It's also inline with where JS is going with standards like `fetch()` and `async/await` that are based on Promises.

See `./examples` to see how to use the new API.

### Breaking changes

- `say` renamed to `send` to reflect that it deals with more than just text
- Removed built-in actions `merge` and `error`
- Actions signature simplified with `request` and `response` arguments
- Actions need to return promises and do not receive the `cb` parameter anymore
- INFO level replaces LOG level
- configuration is now done when instantiating the `Wit` object, instead of using env vars

## v3.3.2

- allows for overriding API version, by setting `WIT_API_VERSION`

## v3.3.1

- adding API versioning (defaults to `20160516`)
- warns instead of throwing when validating actions
- fixing null values when cloning context

## v3.3.0

- callbacks are not called asynchronously by default, choice is left to the developer (use process.nextTick in your callback to emulate the previous behavior)
- using `node-fetch` instead of `requests`
- the `message()` API takes now an optional context as second parameter

## v3.2.2

- fixing context not updated in interactive mode
- fixing array values in context
- create readline interface only in interactive mode

## v3.2.0

Unifying action parameters.

### breaking

- the `say` action now takes 4 parameters: `sessionId`, `context`, `message`, `cb`
- the `error` action now takes 3 parameters: `sessionId`, `context`, `error`

## v3.1.0

Updating action parameters.

### breaking

- the `merge` action now takes 5 parameters: `sessionId`, `context`, `entities`, `message`, `cb`
- the `error` action now takes the context as second parameter
- custom actions now take 3 parameters: `sessionId`, `context`, `cb`

## v3.0.0

Bot Engine integration

### breaking

- the library now provides a Wit object
- `captureTextIntent` has been moved to `Wit.message` with no token
- audio not supported


lib/wit.js
/**
 * Copyright (c) Meta Platforms, Inc. and its affiliates. All Rights Reserved.
 */

'use strict';

const {DEFAULT_API_VERSION, DEFAULT_WIT_URL} = require('./config');
const log = require('./log');
const fetch = require('isomorphic-fetch');
const EventEmitter = require('events');
const HttpsProxyAgent = require('https-proxy-agent');
const {Readable} = require('stream');
const Url = require('url');

const LIVE_UNDERSTANDING_API_VERSION = 20220608;
const MULTI_RESPONSES_API_VERSION = 20230215;

class Wit extends EventEmitter {
  constructor(opts) {
    super();
    this.config = Object.freeze(validate(opts));
  }

  runComposer(sessionId, contextMap, message) {
    return this.event(sessionId, contextMap, message).then(
      this.makeComposerHandler(sessionId),
    );
  }

  runComposerAudio(sessionId, contentType, body, contextMap) {
    return this.converse(sessionId, contentType, body, contextMap).then(
      this.makeComposerHandler(sessionId),
    );
  }

  converse(sessionId, contentType, body, contextMap) {
    if (typeof sessionId !== 'string') {
      throw new Error('Please provide a session ID (string).');
    }

    if (typeof contentType !== 'string') {
      throw new Error('Please provide a content-type (string).');
    }

    if (!body instanceof Readable) {
      throw new Error('Please provide an audio stream (Readable).');
    }

    const {actions, apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      session_id: sessionId,
      v: apiVersion,
    };

    if (typeof contextMap === 'object') {
      params.context_map = JSON.stringify(contextMap);
    }

    const method = 'POST';
    const fullURL = witURL + '/converse?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    const multi_responses_enabled = apiVersion >= MULTI_RESPONSES_API_VERSION;

    const req = fetch(fullURL, {
      body,
      method,
      proxy,
      headers: {
        ...headers,
        'Content-Type': contentType,
        'Transfer-Encoding': 'chunked',
      },
    });

    const _partialResponses = req
      .then(
        resp =>
          new Promise((resolve, reject) => {
            logger.debug('status', resp.status);
            const bodyStream = resp.body;

            bodyStream.on('readable', () => {
              let chunk;
              let contents = '';
              while (null !== (chunk = bodyStream.read())) {
                contents += chunk.toString();
              }

              for (const rsp of parseResponse(contents)) {
                const {action, context_map, error, is_final, response, text} =
                  rsp;

                // Live transcription
                if (!(error || is_final)) {
                  logger.debug('[converse] partialTranscription:', text);
                  this.emit('partialTranscription', text);
                }

                // Multi-responses
                if (
                  multi_responses_enabled &&
                  !(error || is_final) &&
                  (action || response)
                ) {
                  if (response) {
                    logger.debug('[converse] partialResponse:', response);
                    this.emit('response', response);
                  }

                  if (action) {
                    logger.debug('[converse] got partial action:', action);
                    runAction(logger, actions, action, context_map);
                  }
                }
              }
            });
          }),
      )
      .catch(e =>
        logger.error('[converse] could not parse partial response', e),
      );

    return req
      .then(response => Promise.all([response.text(), response.status]))
      .then(([contents, status]) => {
        const finalResponse = parseResponse(contents).pop();
        const {text} = finalResponse;

        logger.debug('[converse] fullTranscription:', text);
        this.emit('fullTranscription', text);

        return [finalResponse, status];
      })
      .catch(e => e)
      .then(makeWitResponseHandler(logger, 'converse'));
  }

  event(sessionId, contextMap, message) {
    if (typeof sessionId !== 'string') {
      throw new Error('Please provide a session ID (string).');
    }

    const {actions, apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      session_id: sessionId,
      v: apiVersion,
    };

    if (typeof contextMap === 'object') {
      params.context_map = JSON.stringify(contextMap);
    }

    const body = {};
    if (typeof message === 'string') {
      body.type = 'message';
      body.message = message;
    }

    const method = 'POST';
    const fullURL = witURL + '/event?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    const req = fetch(fullURL, {
      body: JSON.stringify(body),
      method,
      headers,
      proxy,
    });

    // Multi-responses
    if (apiVersion >= MULTI_RESPONSES_API_VERSION) {
      const _partialResponses = req
        .then(
          resp =>
            new Promise((resolve, reject) => {
              logger.debug('status', resp.status);
              const bodyStream = resp.body;

              bodyStream.on('readable', () => {
                let chunk;
                let contents = '';
                while (null !== (chunk = bodyStream.read())) {
                  contents += chunk.toString();
                }

                for (const rsp of parseResponse(contents)) {
                  const {action, context_map, error, is_final, response} = rsp;

                  if (!(error || is_final) && (action || response)) {
                    if (response) {
                      logger.debug('[event] partialResponse:', response);
                      this.emit('response', response);
                    }

                    if (action) {
                      logger.debug('[event] got partial action:', action);
                      runAction(logger, actions, action, context_map);
                    }
                  }
                }
              });
            }),
        )
        .catch(e =>
          logger.error('[event] could not parse partial response', e),
        );
    }

    return req
      .then(response => Promise.all([response.text(), response.status]))
      .then(([contents, status]) => ([parseResponse(contents).pop(), status]))
      .catch(e => e)
      .then(makeWitResponseHandler(logger, 'event'));
  }

  message(q, context, n) {
    if (typeof q !== 'string') {
      throw new Error('Please provide a text input (string).');
    }

    const {apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      q,
      v: apiVersion,
    };

    if (typeof context === 'object') {
      params.context = JSON.stringify(context);
    }

    if (typeof n === 'number') {
      params.n = JSON.stringify(n);
    }

    const method = 'GET';
    const fullURL = witURL + '/message?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    return fetch(fullURL, {
      method,
      headers,
      proxy,
    })
      .then(response => Promise.all([response.json(), response.status]))
      .then(makeWitResponseHandler(logger, 'message'));
  }

  speech(contentType, body, context, n) {
    if (typeof contentType !== 'string') {
      throw new Error('Please provide a content-type (string).');
    }

    if (!body instanceof Readable) {
      throw new Error('Please provide an audio stream (Readable).');
    }

    const {apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      v: apiVersion,
    };

    if (typeof context === 'object') {
      params.context = JSON.stringify(context);
    }

    if (typeof n === 'number') {
      params.n = JSON.stringify(n);
    }

    const method = 'POST';
    const fullURL = witURL + '/speech?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    const live_understanding_enabled =
      apiVersion >= LIVE_UNDERSTANDING_API_VERSION;

    const req = fetch(fullURL, {
      body,
      method,
      agent: proxy,
      headers: {
        ...headers,
        'Content-Type': contentType,
        'Transfer-Encoding': 'chunked',
      },
    });

    const _partialResponses = req
      .then(
        response =>
          new Promise((resolve, reject) => {
            logger.debug('status', response.status);
            const bodyStream = response.body;

            bodyStream.on('readable', () => {
              let chunk;
              let contents = '';
              while (null !== (chunk = bodyStream.read())) {
                contents += chunk.toString();
              }

              for (const rsp of parseResponse(contents)) {
                const {error, intents, is_final, text} = rsp;

                // Live transcription
                if (!(error || intents)) {
                  logger.debug('[speech] partialTranscription:', text);
                  this.emit('partialTranscription', text);
                }

                // Live understanding
                if (live_understanding_enabled && intents && !is_final) {
                  logger.debug('[speech] partialUnderstanding:', rsp);
                  this.emit('partialUnderstanding', rsp);
                }
              }
            });
          }),
      )
      .catch(e => logger.error('[speech] could not parse partial response', e));

    return req
      .then(response => Promise.all([response.text(), response.status]))
      .then(([contents, status]) => {
        const finalResponse = parseResponse(contents).pop();
        const {text} = finalResponse;

        logger.debug('[speech] fullTranscription:', text);
        this.emit('fullTranscription', text);

        return [finalResponse, status];
      })
      .catch(e => e)
      .then(makeWitResponseHandler(logger, 'speech'));
  }

  dictation(contentType, body) {
    if (typeof contentType !== 'string') {
      throw new Error('Please provide a content-type (string).');
    }

    if (!body instanceof Readable) {
      throw new Error('Please provide an audio stream (Readable).');
    }

    const {apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      v: apiVersion,
    };

    const method = 'POST';
    const fullURL = witURL + '/dictation?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    const req = fetch(fullURL, {
      body,
      method,
      proxy,
      headers: {
        ...headers,
        'Content-Type': contentType,
        'Transfer-Encoding': 'chunked',
      },
    });

    const _partialResponses = req
      .then(
        response =>
          new Promise((resolve, reject) => {
            logger.debug('status', response.status);
            const bodyStream = response.body;
            bodyStream.on('readable', () => {
              let chunk;
              let contents = '';
              while (null !== (chunk = bodyStream.read())) {
                contents += chunk.toString();
              }

              for (const rsp of parseResponse(contents)) {
                const {error, is_final, text} = rsp;

                // Live transcription
                if (!error) {
                  if (!is_final) {
                    logger.debug('[dictation] partial transcription:', text);
                    this.emit('partialTranscription', text);
                  } else {
                    logger.debug(
                      '[dictation] full sentence transcription:',
                      text,
                    );
                    this.emit('fullTranscription', text);
                  }
                }
              }
            });
          }),
      )
      .catch(e =>
        logger.error('[dictation] could not parse partial response', e),
      );

    return req
      .then(response => Promise.all([response.text(), response.status]))
      .then(([contents, status]) => {
        const finalResponse = parseResponse(contents).pop();
        const {text} = finalResponse;

        logger.debug('[dictation] last full sentence transcription:', text);
        this.emit('fullTranscription', text);

        return [finalResponse, status];
      })
      .catch(e => e)
      .then(makeWitResponseHandler(logger, 'dictation'));
  }

  synthesize(
    q,
    voice,
    style = 'default',
    speed = 100,
    pitch = 100,
    gain = 100,
  ) {
    if (typeof q !== 'string') {
      throw new Error('Please provide a text input (string).');
    }
    if (typeof voice !== 'string') {
      throw new Error('Please provide a voice input (string).');
    }

    const {apiVersion, headers, logger, proxy, witURL} = this.config;

    const params = {
      v: apiVersion,
    };

    const body = {
      q: q,
      voice: voice,
      style: style,
      speed: speed,
      pitch: pitch,
      gain: gain,
    };

    const method = 'POST';
    const fullURL = witURL + '/synthesize?' + encodeURIParams(params);
    logger.debug(method, fullURL);

    return fetch(fullURL, {
      body: JSON.stringify(body),
      method,
      proxy,
      headers: {
        ...headers,
        'Content-Type': 'application/json',
      },
    })
      .then(response => Promise.all([response, response.status]))
      .then(makeWitResponseHandler(logger, 'synthesize'));
  }

  makeComposerHandler(sessionId) {
    const {actions, logger} = this.config;

    return ({context_map, action, expects_input, response}) => {
      if (typeof context_map !== 'object') {
        throw new Error(
          'Unexpected context_map in API response: ' +
            JSON.stringify(context_map),
        );
      }

      if (response) {
        logger.debug('[composer] response:', response);
        this.emit('response', response);
      }

      if (action) {
        logger.debug('[composer] got action', action);
        return runAction(logger, actions, action, context_map).then(
          ({context_map, stop}) => {
            if (expects_input && !stop) {
              return this.runComposer(sessionId, context_map);
            }
            return {context_map};
          },
        );
      }

      return {context_map, expects_input};
    };
  }
}

const runAction = (logger, actions, name, ...rest) => {
  logger.debug('Running action', name);
  return Promise.resolve(actions[name](...rest));
};

const makeWitResponseHandler = (logger, endpoint) => rsp => {
  const error = e => {
    logger.error('[' + endpoint + '] Error: ' + e);
    throw e;
  };

  if (rsp instanceof Error) {
    return error(rsp);
  }

  const [json, status] = rsp;

  if (json instanceof Error) {
    return error(json);
  }

  const err = json.error || (status !== 200 && json.body + ' (' + status + ')');

  if (err) {
    return error(err);
  }

  logger.debug('[' + endpoint + '] Response: ' + JSON.stringify(json));
  return json;
};

const getProxyAgent = witURL => {
  const url = Url.parse(witURL);
  const proxy =
    url.protocol === 'http:'
      ? process.env.http_proxy || process.env.HTTP_PROXY
      : process.env.https_proxy || process.env.HTTPS_PROXY;
  const noProxy = process.env.no_proxy || process.env.NO_PROXY;

  const shouldIgnore = noProxy && noProxy.indexOf(url.hostname) > -1;
  if (proxy && !shouldIgnore) {
    return new HttpsProxyAgent(proxy);
  }

  if (!proxy) {
    return null;
  }
};

const encodeURIParams = params =>
  Object.entries(params)
    .map(([key, value]) => key + '=' + encodeURIComponent(value))
    .join('&');

const parseResponse = response => {
  const chunks = response
    .split('\r\n')
    .map(x => x.trim())
    .filter(x => x.length > 0);

  let prev = '';
  let jsons = [];
  for (const chunk of chunks) {
    try {
      prev += chunk;
      jsons.push(JSON.parse(prev));
      prev = '';
    } catch (_e) {}
  }

  return jsons;
};

const validate = opts => {
  if (!opts.accessToken) {
    throw new Error(
      'Could not find access token, learn more at https://wit.ai/docs',
    );
  }

  opts.witURL = opts.witURL || DEFAULT_WIT_URL;
  opts.apiVersion = opts.apiVersion || DEFAULT_API_VERSION;
  opts.headers = opts.headers || {
    Authorization: 'Bearer ' + opts.accessToken,
    'Content-Type': 'application/json',
  };
  opts.logger = opts.logger || new log.Logger(log.INFO);
  opts.proxy = getProxyAgent(opts.witURL);

  if (opts.actions && typeof opts.actions !== 'object') {
    throw new Error('Please provide actions mapping (string -> function).');
  }

  return opts;
};

module.exports = Wit;
